DOCS-#2996: document `modin.backends.base`

dchigarev opened the pull request:
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

<!--
Thank you for your contribution!
Please review the contributing docs: https://modin.readthedocs.io/en/latest/CONTRIBUTING.html
if you have questions about contributing.
-->

## What do these changes do?

<!-- Please give a short brief about these changes. -->

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #2996 <!-- issue must be created for each patch -->
- [x] tests added and passing
- This PR adds doc-strings for all of the `BaseQueryCompiler` methods and so defines common query compilers API. PR appears to be really huge but it's required to finally define the first version of backends API which does not depend on pandas (however many function params still don't have any description and so refers to pandas documentation).
- The next step of setting up common query compilers API is to get rid of `*args` and `**kwargs` which hides actual method parameters. In this PR actual method parameters were inferred from high-level API calls signatures and placed into the doc-strings.

Built docs for this PR: https://modin--3021.org.readthedocs.build/en/3021/index.html
Affected sections:
- https://modin--3021.org.readthedocs.build/en/3021/flow/modin/backends/
- https://modin--3021.org.readthedocs.build/en/3021/flow/modin/backends/base/query_compiler.html

dchigarev commented:
@vnlitvinov @YarShev @anmyachev the flow docs are ready for review. Docstrings are in progress

amyskov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1702,218 +3478,1392 @@ def set_index_names(self, names, axis=0):
+    # which we shouldn't do at the backend. We need to move this at the front.
+    def resample_pipe(self, resample_args, func, *args, **kwargs):
+        """
+        Resample time-series data and apply aggregation on it.
```
Need to add something about piping themself.

dchigarev replied:
What do you mean under "pipping themself"? Current doc-string represents what this function is: "apply passed function to the `pandas.Resampler` object".

Added reference to `Resampler.pipe` method.

amyskov replied:
`pandas.Resampler.pipe` allows to call `pipe` multiple times for calls chaining. If there is no usage difference with pandas the reference will be enough.

amyskov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1702,218 +3478,1392 @@ def set_index_names(self, names, axis=0):
+        kwargs : dict, optional""",
+        build_rulles="udf_aggregation",
+    )
+    def rolling_apply(
```
Need to highlight the difference to `rolling_aggregate` in the summary.

dchigarev replied:
There is no much difference to highlight. At the query compiler level `apply` is a special case of `aggregate` (`apply` takes one function to execute, when `aggregate` can take multiple functions). `apply` leaked at the backend level by mistake, because the parameters it provides in addition to `aggregate` make no sense at the query compiler. Added deprecation warning at the doc-string for this function.

amyskov reviewed (commented):
Great work! left some minor comments.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -239,216 +1076,455 @@ def to_numpy(self, **kwargs):
+        ----------
+        other : BaseQueryCompiler
+            Frame to grab replacement values from.
+        join : {"left"}
```
Yes, going from [pandas docs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.update.html) there is only one join type. An actual "reference" implementation from pandas query compiler also supports only one type of join:
https://github.com/modin-project/modin/blob/626446e6999fde43ff32cbee5df1b1c147d6d7b2/modin/backends/pandas/query_compiler.py#L347-L350

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -26,13 +34,739 @@
+        Sign which represents specified binary operation.
+    r : bool, default: False
+        Whether `self` is the right operand.
+    op_type : {"arithmetic", "logical", "comparison"}, default: arithmetic
```
added quotes to the default value

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -51,30 +798,43 @@ def axis_setter(self, labels):
-    Note: See the Abstract Methods and Fields section immediately below this
-        for a list of requirements for subclassing this object.
+    """
+    Abstract Class that handles the queries to Modin dataframes.
```
fixed

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -105,24 +895,54 @@ def add_suffix(self, suffix, axis=1):
 
-    def concat(self, axis, other, **kwargs):
-        """Concatenates two objects together.
+    def concat(self, axis, other, **kwargs):  # noqa: PR02
```
added `FIXME` comment at the top of the `BaseQueryCompiler` definition

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -239,216 +1076,455 @@ def to_numpy(self, **kwargs):
+    @add_refer_to("DataFrame.where")
+    def where(self, cond, other, **kwargs):  # noqa: PR02
+        """
+        Update values of self using values from `other` at positions where `cond` is True.
```
my typo, fixed that

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1074,6 +2791,52 @@ def apply(self, func, axis, *args, **kwargs):
+    # FIXME: `map_args` and `reduce_args` leaked there from `PandasQueryCompiler.groupby_*`,
+    # pandas backend implements groupby via MapReduce approach, but for other backends these
+    # parameters make no sense, they shouldn't be presented in a base class.
+    __doc_groupby_method = """
```
moved

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -524,154 +1643,411 @@ def set_index_from_columns(
+        ----------
+        to_replace : scalar, list-like, regex, modin.pandas.Series, or None
+        value : scalar, list-like, regex or dict
+        inplace : False
```
`inplace` parameters takes only a single value, changed its doc to: `inplace : {False}`

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -169,29 +989,33 @@ def finalize(self):
             The pandas DataFrame to convert from.
-        data_cls :
-            Modin DataFrame object to convert to.
+        data_cls : cls
```
How can we represent the type of classes, which are inherited from `BasePandasFrame`?

amyskov replied:
I think you are right, `cls` fits better here, let's leave it as is.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1673,9 +3449,9 @@ def set_index_name(self, name, axis=0):
+        name : hashable
             New index name.
-        axis: int (default 0),
+        axis : {0, 1}, default: 0
```
aligned all `axis` parameter descriptions to this format

amyskov commented:
LGTM!

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+
+Brief description
+'''''''''''''''''
+``BaseQueryCompiler`` is an abstract class of query compiler, it sets a common interface,
```
```suggestion
``BaseQueryCompiler`` is an abstract class of query compiler, and sets a common interface
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+Brief description
+'''''''''''''''''
+``BaseQueryCompiler`` is an abstract class of query compiler, it sets a common interface,
+that every other query compilers in Modin have to follow. Base class contains basic
```
```suggestion
that every other query compiler implementation in Modin must follow. The Base class contains a basic
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+'''''''''''''''''
+``BaseQueryCompiler`` is an abstract class of query compiler, it sets a common interface,
+that every other query compilers in Modin have to follow. Base class contains basic
+implementations for most of the interface methods, all them are
```
```suggestion
implementations for most of the interface methods, all of which
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+``BaseQueryCompiler`` is an abstract class of query compiler, it sets a common interface,
+that every other query compilers in Modin have to follow. Base class contains basic
+implementations for most of the interface methods, all them are
+:ref:`defaulting to pandas <defaulting-to-pandas-mechanism>`.
```
```suggestion
:ref:`default to pandas <defaulting-to-pandas-mechanism>`.
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+
+Subclassing ``BaseQueryCompiler``
+'''''''''''''''''''''''''''''''''
+If you want to add new type of query compiler to Modin you need to inherit new
```
```suggestion
If you want to add new type of query compiler to Modin the new class needs to inherit
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+Subclassing ``BaseQueryCompiler``
+'''''''''''''''''''''''''''''''''
+If you want to add new type of query compiler to Modin you need to inherit new
+class from `BaseQueryCompiler` and implement the rest of the abstract methods:
```
```suggestion
from `BaseQueryCompiler` and implement the abstract methods:
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+- ``finalize`` finalize object constructing.
+- ``free`` trigger memory cleaning.
+
+(Please refer to the code documentation to see full contracts for these functions).
```
```suggestion
(Please refer to the code documentation to see the full documentation for these functions).
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+
+(Please refer to the code documentation to see full contracts for these functions).
+
+This is a minimum set of operations to ensure the proper work of your new query compiler,
```
```suggestion
This is a minimum set of operations to ensure a new query compiler will function in the Modin architecture,
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+(Please refer to the code documentation to see full contracts for these functions).
+
+This is a minimum set of operations to ensure the proper work of your new query compiler,
+the rest of the API will be defaulted pandas. To add backend-specific implementation for
```
```suggestion
and the rest of the API can safely default to the pandas implementation. To add a backend-specific implementation for
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+
+This is a minimum set of operations to ensure the proper work of your new query compiler,
+the rest of the API will be defaulted pandas. To add backend-specific implementation for
+some of the query compiler operation, just override the corresponding method in your
```
```suggestion
some of the query compiler operations, just override the corresponding method in your
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+Example
+'''''''
+As an exercise let's define a new query compiler in `Modin`, just to see how easy it is.
+Usually, query compiler routes formed queries to the underlying :doc:`frame </flow/modin/engines/base/frame/data>` class,
```
```suggestion
Usually, the query compiler routes formed queries to the underlying :doc:`frame </flow/modin/engines/base/frame/data>` class,
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+'''''''
+As an exercise let's define a new query compiler in `Modin`, just to see how easy it is.
+Usually, query compiler routes formed queries to the underlying :doc:`frame </flow/modin/engines/base/frame/data>` class,
+which represents an actual execution engine and responsible for executing queries. In the glory
```
```suggestion
which represents submits operators to an execution engine. For the sake
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,93 @@
+of simplicity and independence of this example, our execution engine will be the `pandas` itself.
+
+So, we need to inherit a new class from ``BaseQueryCompiler`` and implement all of the abstract methods.
+In case of `pandas` as an execution engine it's a trivial task:
```
```suggestion
In this case, with `pandas` as an execution engine, it's trivial:
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+==============
+
+Modin has several execution backends. Calling any DataFrame API function will end up in
+some backend-specific method. Query compiler is a bridge between Modin DataFrame and
```
```suggestion
some backend-specific method. The query compiler is a bridge between Modin Dataframe and
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+Query compiler
+==============
+
+Modin has several execution backends. Calling any DataFrame API function will end up in
```
```suggestion
Modin supports several execution backends. Calling any DataFrame API function will end up in
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+    :align: right
+    :width: 300px
+
+Query compilers of all backends provide a common API, which is used by Modin DataFrame
```
```suggestion
Query compilers of all backends implement a common API, which is used by the Modin Dataframe
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+    :width: 300px
+
+Query compilers of all backends provide a common API, which is used by Modin DataFrame
+to emulate pandas API. The role of the query compiler is to translate its API into
```
```suggestion
to support dataframe queries. The role of the query compiler is to translate its API into
```

dchigarev replied:
Does DataFrame uses a query compiler to support "dataframe queries"? What are "dataframe queries" and why we should use this term instead of "pandas API"?

devin-petersohn replied:
The Modin implementation is more general than just pandas. Since this is developer documentation, we can assume a higher technical level and talk more generally.

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+
+Query compilers of all backends provide a common API, which is used by Modin DataFrame
+to emulate pandas API. The role of the query compiler is to translate its API into
+a set of DataFrame algebra operations. Each query compiler instance contains
```
```suggestion
a pairing of known user-defined functions and dataframe algebra operators. Each query compiler instance contains a
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+Query compilers of all backends provide a common API, which is used by Modin DataFrame
+to emulate pandas API. The role of the query compiler is to translate its API into
+a set of DataFrame algebra operations. Each query compiler instance contains
+:doc:`frame </flow/modin/engines/base/frame/index>` of the selected execution engine and query
```
```suggestion
:doc:`frame </flow/modin/engines/base/frame/index>` of the selected execution engine and queries
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+it with the compiled queries to get the result. The query compiler object is immutable,
+so the result of every method is a new query compiler.
+
+Query compilers API is defined by the :doc:`BaseQueryCompiler <base/query_compiler>` class
```
```suggestion
The query compilers API is defined by the :doc:`BaseQueryCompiler <base/query_compiler>` class
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+so the result of every method is a new query compiler.
+
+Query compilers API is defined by the :doc:`BaseQueryCompiler <base/query_compiler>` class
+and partially mimic the pandas API, however, they're not equal. Query compilers API
```
```suggestion
and may resemble the pandas API, however, they're not equal. The query compilers API
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+Query compilers API is defined by the :doc:`BaseQueryCompiler <base/query_compiler>` class
+and partially mimic the pandas API, however, they're not equal. Query compilers API
+is significantly reduced in comparison with pandas, since many corner cases or even the
+whole methods can be handled at the DataFrame level with the existing API.
```
```suggestion
whole methods can be handled at the API layer with the existing API.
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+is significantly reduced in comparison with pandas, since many corner cases or even the
+whole methods can be handled at the DataFrame level with the existing API.
+
+Query compiler is the level where Modin stops distinguishing Frame and Series objects.
```
```suggestion
The query compiler is the level where Modin stops distinguishing Frame and Series (or column) objects.
```

dchigarev replied:
just a bit of curiosity, what does "Series **or column**" means?

YarShev replied:
I believe @devin-petersohn mean that we treat Series (or column - column of a DataFrame is a Series object as well) as DataFrame at the QC layer.

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+whole methods can be handled at the DataFrame level with the existing API.
+
+Query compiler is the level where Modin stops distinguishing Frame and Series objects.
+Series is represented by one-column query compiler, where Series name is the column label,
```
```suggestion
A Series is represented by a one-column query compiler, where the Series name is the column label.
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+
+Query compiler is the level where Modin stops distinguishing Frame and Series objects.
+Series is represented by one-column query compiler, where Series name is the column label,
+if Series is unnamed, then the label would be ``"__reduced__"``. DataFrame API level
```
```suggestion
If Series is unnamed, then the label is ``"__reduced__"``. The Dataframe API layer
```

devin-petersohn commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,46 @@
+Query compiler is the level where Modin stops distinguishing Frame and Series objects.
+Series is represented by one-column query compiler, where Series name is the column label,
+if Series is unnamed, then the label would be ``"__reduced__"``. DataFrame API level
+interprets one-column query compilers as Series or DataFrame depending on the operation context.
```
```suggestion
interprets a one-column query compiler as Series or DataFrame depending on the operation context.
```

devin-petersohn reviewed (commented):
Thanks @dchigarev, left a few comments.

prutskov reviewed (commented):
General comments:

1. I prefer to use: one backticks for highlighting parameters of func/method(by rules of numpydoc), double backticks for highlighting code-entities as funcs/classes (In sphinx built of documentation this will be immediately visible in the text)
2. Why we don't add .rst in any `toctree`?
3. Are you plan to add API section in .rst files to make possible auto-links to classes/functions in .rst docs?

dchigarev commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,91 @@
+Base query compiler
```
included

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        Returns
+        -------
+        BaseQueryCompiler
+            `QueryCompiler` of the same shape as `self`, where each element is the {method}
```
It's neither code entity nor function parameter, it's just a term such as DataFrame. So removed backticks at all.

dchigarev commented:
@prutskov
> Are you plan to add API section in .rst files to make possible auto-links to classes/functions in .rst docs?

I'm not sure where to place it. For now, I've added query compiler API to the [`BaseQueryCompiler` page](https://modin--3021.org.readthedocs.build/en/3021/flow/modin/backends/base/query_compiler.html#query-compiler-api) just to check that it's generating properly. But I don't think that the flow documentation is a suitable place for the code docs. IMO there should be a separate API section where all of the code documentation will be stored (just like in pandas [[pandas API section]](https://pandas.pydata.org/docs/reference/index.html)) so that if you want to read about the execution flow you go to the flow docs, and if you're interested in the API, you go to the API section.

@vnlitvinov are there any plans to detach all of the generated API docs (`modin/engines` API, `sklearn` API) into a separate section?

vnlitvinov commented:
> @vnlitvinov are there any plans to detach all of the generated API docs (`modin/engines` API, `sklearn` API) into a separate section?

Not now. Current plans are to document most things, then give a thorough re-assessment of what we have and then re-structure the documentation. But we cannot restructure it until we have it done :)

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+
+Brief description
+'''''''''''''''''
+``BaseQueryCompiler`` is an abstract class of query compiler, and sets a common interface
```
```suggestion
:py:class:`~modin.backends.base.query_compiler.BaseQueryCompiler` is an abstract class of query compiler, and sets a common interface
```
This makes auto-link to `BaseQueryCompiler` class API

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+implementations for most of the interface methods, all of which
+:ref:`default to pandas <defaulting-to-pandas-mechanism>`.
+
+Subclassing ``BaseQueryCompiler``
```
```suggestion
Subclassing :py:class:`~modin.backends.base.query_compiler.BaseQueryCompiler`
```

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+Subclassing ``BaseQueryCompiler``
+'''''''''''''''''''''''''''''''''
+If you want to add new type of query compiler to Modin the new class needs to inherit
+from ``BaseQueryCompiler`` and implement the abstract methods:
```
```suggestion
from :py:class:`~modin.backends.base.query_compiler.BaseQueryCompiler` and implement the abstract methods:
```

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+which represents submits operators to an execution engine. For the sake
+of simplicity and independence of this example, our execution engine will be the `pandas` itself.
+
+So, we need to inherit a new class from ``BaseQueryCompiler`` and implement all of the abstract methods.
```
```suggestion
So, we need to inherit a new class from :py:class:`~modin.backends.base.query_compiler.BaseQueryCompiler` and implement all of the abstract methods.
```

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+If you want to add new type of query compiler to Modin the new class needs to inherit
+from ``BaseQueryCompiler`` and implement the abstract methods:
+
+- ``from_pandas`` build query compiler from pandas DataFrame.
```
```suggestion
- :py:meth:`~modin.backends.base.query_compiler.BaseQueryCompiler.from_pandas` build query compiler from pandas DataFrame.
```

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+from ``BaseQueryCompiler`` and implement the abstract methods:
+
+- ``from_pandas`` build query compiler from pandas DataFrame.
+- ``from_arrow`` build query compiler from Arrow Table.
```
```suggestion
- :py:meth:`~modin.backends.base.query_compiler.BaseQueryCompiler.from_arrow` build query compiler from Arrow Table.
```

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+
+- ``from_pandas`` build query compiler from pandas DataFrame.
+- ``from_arrow`` build query compiler from Arrow Table.
+- ``to_pandas`` get query compilers representation via pandas DataFrame.
```
```suggestion
- :py:meth:`~modin.backends.base.query_compiler.BaseQueryCompiler.to_pandas` get query compilers representation via pandas DataFrame.
```

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+- ``from_pandas`` build query compiler from pandas DataFrame.
+- ``from_arrow`` build query compiler from Arrow Table.
+- ``to_pandas`` get query compilers representation via pandas DataFrame.
+- ``default_to_pandas`` do :ref:`fallback to pandas <defaulting-to-pandas-mechanism>` for the passed function. 
```
```suggestion
- :py:meth:`~modin.backends.base.query_compiler.BaseQueryCompiler.default_to_pandas` do :ref:`fallback to pandas <defaulting-to-pandas-mechanism>` for the passed function. 
```

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+- ``from_arrow`` build query compiler from Arrow Table.
+- ``to_pandas`` get query compilers representation via pandas DataFrame.
+- ``default_to_pandas`` do :ref:`fallback to pandas <defaulting-to-pandas-mechanism>` for the passed function. 
+- ``finalize`` finalize object constructing.
```
```suggestion
- :py:meth:`~modin.backends.base.query_compiler.BaseQueryCompiler.finalize` finalize object constructing.
```

prutskov commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+- ``to_pandas`` get query compilers representation via pandas DataFrame.
+- ``default_to_pandas`` do :ref:`fallback to pandas <defaulting-to-pandas-mechanism>` for the passed function. 
+- ``finalize`` finalize object constructing.
+- ``free`` trigger memory cleaning.
```
```suggestion
- :py:meth:`~modin.backends.base.query_compiler.BaseQueryCompiler.free` trigger memory cleaning.
```

prutskov commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,52 @@
+it with the compiled queries to get the result. The query compiler object is immutable,
+so the result of every method is a new query compiler.
+
+The query compilers API is defined by the :doc:`BaseQueryCompiler <base/query_compiler>` class
```
```suggestion
The query compilers API is defined by the :py:class:`~modin.backends.base.query_compiler.BaseQueryCompiler` class
```
It generates auto-link to `BaseQueryCompiler` class API in other .rst file

prutskov commented on docs/flow/modin/backends/index.rst:
```diff
@@ -0,0 +1,52 @@
+
+- :doc:`Base module<base/query_compiler>` contains an abstract query compiler class which defines common API.
+- :doc:`Pandas module<pandas/index>` contains query compiler and text parsers for pandas backend.
+- :doc:`Cudf module<cudf/index>` contains query compiler and text parsers for Cudf backend.
```
Suggestion above wasn't applied, but resolved. Can you explain?

dchigarev replied:
Can you please refer to that suggestion? I thought that I applied all of them

prutskov replied:
It's here https://github.com/modin-project/modin/pull/3021#discussion_r631825296

dchigarev replied:
Oh, thanks. Fixed that in the last commit as well.

prutskov reviewed (commented):
Mostly LGTM.

In my opinion, we can add API section in flow .rst files, in the future we can move all API docs to separate folder as you propose.
I propose to use links on entities to give fast access to API by 'clicking'. Also, it makes links independent from the location of generated API docs.

prutskov reviewed (dismissed):
Thanks @dchigarev, LGTM!

prutskov reviewed (approved):
LGTM

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+        **kwargs : dict""",
+        overwrite_template_params=True,
+    )
+    def resample_interpolate(
+        self,
+        resample_args,
+        method,
+        axis,
+        limit,
+        inplace,
+        limit_direction,
+        limit_area,
+        downcast,
+        **kwargs,
+    ):
+        return ResampleDefault.register(pandas.core.resample.Resampler.interpolate)(
+            self,
+            resample_args,
+            method,
+            axis,
+            limit,
+            inplace,
+            limit_direction,
+            limit_area,
+            downcast,
+            **kwargs,
+        )
```
this would certainly benefit in readability from being replaced by
```python
resample_interpolate = _doc_resample_fillna(......)(ResampleDefault.register(pandas.core.resample.Resampler.interpolate))
```

dchigarev replied:
but then we lose the signature of this function, the function would just take `*args` and `**kwargs` and it would be completely unclear what the actual parameters of this method. Function without solid signature more likely to get unsynced with its docs.

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+            self, resample_args, limit
+        )
+
+    # FIXME: This method require us to build high-level resampler object
+    # which we shouldn't do at the backend. We need to move this at the front.
```
I suggest filing an issue about this

dchigarev replied:
created #3105

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+            self, resample_args, ddof, *args, **kwargs
+        )
+
+    @_doc_resample_reduction(result="number of element in a group", link="size")
```
```suggestion
    @_doc_resample_reduction(result="number of elements in a group", link="size")
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+
+    # FIXME: This method require us to build high-level resampler object
+    # which we shouldn't do at the backend. We need to move this at the front.
+    def resample_transform(self, resample_args, arg, *args, **kwargs):
```
what's the difference between `resample_pipe` and `resample_transform`? their docstrings seem to be equal

dchigarev replied:
accidentally copy-pasted its doc-string)
Now `resample_transform` has the proper doc-string

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+            self, pat, flags, **kwargs
+        )
+
+    @_doc_str_method(method="get", params="i : int")
```
`i` could use a description IMHO, not only a type

dchigarev replied:
this one is also has a link to the pandas docs with the description of the parameter

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+    def str_match(self, pat, case=True, flags=0, na=np.NaN):
+        return StrDefault.register(pandas.Series.str.match)(self, pat, case, flags, na)
+
+    @_doc_str_method(method="normalize", params="form : {'NFC', 'NFKC', 'NFD', 'NFKD'}")
```
uh, `form` parameter could use a description of what these cryptic acronyms mean...

dchigarev replied:
doc-string has a link to the pandas docs that explains these scary acronyms

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+
+    # Rolling methods
+
+    # FIXME: most of the rolling/window methods takes *args and **kwargs parameters
+    # which only serves the compatibility with numpy, this behaviour is inherited
```
```suggestion
    # FIXME: most of the rolling/window methods take *args and **kwargs parameters
    # which are only needed for the compatibility with numpy, this behaviour is inherited
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+            self, rolling_args, func, *args, **kwargs
+        )
+
+    # FIXME: at the query compiler method `rolling_apply` is an alias for `rolling_aggregate`,
+    # one of these should be removed.
```
I suggest making an issue about this

dchigarev replied:
Created #3107

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+        params="""axis : {0, 1}
+        skipna : bool
+        level : None, default: None
+            Serves the compatibility purpose. Always have to be None.""",
```
```suggestion
            Serves the compatibility purpose. Always has to be None.""",
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+        )
+
+    @_doc_reduce_agg(
+        method="unibased kurtosis", link="kurt", extra_params=["skipna", "**kwargs"]
```
why it does not describe all the parameters?

dchigarev replied:
it describes all parameters, part of the docs are hidden in the list of default parameters inside [`doc_reduce_agg`](https://github.com/dchigarev/modin/blob/e3f0d9a0c9554acaa1bfabf5c46e11984d36a3e5/modin/backends/base/doc_utils.py#L244)

vnlitvinov reviewed (changes requested):
Please also *not* rebase anymore, re-reading a ~4000 LoC diff is hard 🙃

devin-petersohn reviewed (commented):
@dchigarev In addition to @vnlitvinov comments, please also make sure you rebase.

vnlitvinov commented:
> @dchigarev In addition to @vnlitvinov comments, please also make sure you ~rebase~ clear the conflict.

`git merge master` _should_ be able to fix the conflicts (I hope...)

devin-petersohn commented:
@vnlitvinov Creating a merge commit should in the PR is usually fine as long as it is merged with `upstream/master` and not `origin/master` (upstream being `modin-project` and origin being the fork).

dchigarev commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,91 @@
+Brief description
+'''''''''''''''''
+``BaseQueryCompiler`` is an abstract class of query compiler, it sets a common interface,
+that every other query compilers in Modin have to follow. Base class contains basic
```
Was already changed to the following by Devins suggestion:
"that every other query compiler implementation in Modin must follow"

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+'''''''
+As an exercise let's define a new query compiler in `Modin`, just to see how easy it is.
+Usually, the query compiler routes formed queries to the underlying :doc:`frame </flow/modin/engines/base/frame/data>` class,
+which represents submits operators to an execution engine. For the sake
```
```suggestion
which submits operators to an execution engine. For the sake
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+(Please refer to the code documentation to see the full documentation for these functions).
+
+This is a minimum set of operations to ensure a new query compiler will function in the Modin architecture,
+and the rest of the API can safely default to the pandas implementation. To add a backend-specific implementation for
```
```suggestion
and the rest of the API can safely default to the pandas implementation via the base class implementation. To add a backend-specific implementation for
```

devin-petersohn commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+which represents submits operators to an execution engine. For the sake
+of simplicity and independence of this example, our execution engine will be the `pandas` itself.
+
+So, we need to inherit a new class from :py:class:`~modin.backends.base.query_compiler.BaseQueryCompiler` and implement all of the abstract methods.
```
```suggestion
We need to inherit a new class from :py:class:`~modin.backends.base.query_compiler.BaseQueryCompiler` and implement all of the abstract methods.
```

devin-petersohn reviewed (commented):
Thanks @dchigarev, left some comments.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -239,204 +1170,443 @@ def to_numpy(self, **kwargs):
 
-    def add(self, other, **kwargs):
+    @_doc_binary_method(operation="addition", sign="+")
+    def add(self, other, **kwargs):  # noqa: PR02
```
there is a [note](https://github.com/dchigarev/modin/blob/8d804b7d4a235ca00b680f8e813da6ced93be5a2/modin/backends/base/query_compiler.py#L87) about this at the top of the class. Shortly: the silenced warning indicates that there are some parameters documented in the doc-string that are not presented in the function signature, this happens because `**kwargs` hiding actual arguments that we want to document, the solution is to unpack `**kwargs`.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -239,216 +1168,455 @@ def to_numpy(self, **kwargs):
         right : BaseQueryCompiler
-            The query compiler of the right DataFrame to merge with.
+            `QueryCompiler` of the right frame to merge with.
+        how : {"left", "right", "outer", "inner", "cross"}
+        on : label or list of such
+        left_on : label or list of such
+        right_on : label or list of such
+        left_index : bool
+        right_index : bool
+        sort : bool
+        suffixes : list-like
+        copy : bool
+        indicator : bool or str
+        validate : str
```
the description is identical to the same parameters in pandas, this function has a note about this

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -827,143 +2383,356 @@ def describe(self, **kwargs):
+        Parameters
+        ----------
+        n : int, default: 5
+        columns : list of labels, optional
```
done

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        Whether to append note that this method is for one-column
+        query compilers only.
+    try_insert_params_section : bool, default: False
+        Whether to insert parameters section in `template`.
```
we can just stop using this parameter and always assume that the `params` value does not contain the "Parameters" header. applied this

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+            decorated = refer_to_appender(decorated)
+        if one_column_method:
+            decorated = _add_one_column_warning(decorated)
+        return decorated
```
done

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+    broadcast : bool, default: False
+        If `other` is a one-column query compiler, indicates whether it is a Series or not.
+        Frames and Series have to be processed differently, however we can't distinguish them
+        at the query compiler level, so this parameter is a hint that passed from a high level API.
```
yes, the description is correct:
https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/pandas/dataframe.py#L456-L464

the naming comes from the pandas backend, where the way of partitions broadcasting depends on that [[1](https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/data_management/functions/binary_function.py#L71)].

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+
+    return _doc_qc_method(
+        template,
+        params=params_substitution,
```
Indeed. Applied your approach.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+    extra_params = {
+        "logical": """level : int or label
+        In case of MultiIndex match index values on the passed level.
+    axis : {0, 1}
```
fixed

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        to the docstring template.
+    extra_params : list of str, optional
+        Method parameter names to append to the docstring template. Parameter
+        type and description will be grabbed from `extra_params_map`.
```
this map is defined inside the function, added a note about this

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        params = """axis : {0, 1}
+        level : None, default: None
+            Serves the compatibility purpose. Always have to be None.
+        numeric_only : bool, optional"""
```
The description of these parameters copies the description of the same parameters in pandas. Doc-string contains a link to the corresponding method to get the full info about parameters.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        Parameters
+        ----------
+        axis : {{0, 1}}
+        skipna : bool
```
[answer](https://github.com/modin-project/modin/pull/3021#discussion_r641728969)

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+            following rules: {build_rulles}
+        """
+    if params and params.strip(" \t")[-1] != "\n":
+        params += "\n"
```
forgot to remove the debug code)
removed now.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        What method does with the resampled data.
+    output : str
+        What is the content of column names in the result.
+    params : str, optional
```
fixed, now `params` actually optional

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+    -------
+    callable
+    """
+    return _doc_dt(property, "date-time", method, params)
```
you right, the correct spelling would be "datetime".

used `partial` where possible

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        Parameters
+        ----------
+        freq : str
+        ambiguous : "infer", "NaT", bool mask, default: "raise"
```
added "raise" in a list of possible values

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        "rolling window": "rolling_args",
+        "default": "window_args",
+    }
+    window_args_name = window_args_dict.get(win_type, window_args_dict["default"])
```
got rid of using a dictionary in this case

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+    Parameters
+    ----------
+    result : str
+        The result of aggregation.
```
renamed "aggregation" -> "reduction" everywhere at this method

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        0 is for index, when 1 is for columns.
+    groupby_args : dict
+        GroupBy parameters in the format of ``modin.pandas.DataFrame.groupby`` signature.
+    map_args : dict
```
if it isn't implemented as MapReduce then `map_args` will be passed to the only one reduction function. Explained this in the docs.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,829 @@
+        Whether or not to drop non-numeric columns before executing GroupBy.
+    drop : bool, default: False
+        If `by` is a `QueryCompiler` indicates whether or not by-data came
+        from the `self`.
```
The description is correct:
https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/pandas/dataframe.py#L369
The name of this parameter is inherited from the API level, where it means whether or not to drop `by` columns before operation:
https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/pandas/groupby.py#L870-L873

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -239,216 +1168,455 @@ def to_numpy(self, **kwargs):
+        other : BaseQueryCompiler
+            Frame to grab replacement values from.
+        join : {"left"}
+            Specify type of join to allign frames if axes are not equal.
```
added the following note:
"note: currently only one type of join is implemented"
There is no particular reason why the other types of join are not implemented.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -462,53 +1630,96 @@ def columnarize(self):
+            Index-labels to align with.
+        method : {None, "backfill"/"bfill", "pad"/"ffill", "nearest"}
+            Method to use for filling holes in reindexed frame.
+            Please refer to ``modin.pandas.DataFrame.reindex`` for more information.
```
right, removed this

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -524,285 +1735,616 @@ def set_index_from_columns(
 
-    def isin(self, **kwargs):
+    # FIXME:
+    #   1. High-level objects leaks to the query compiler level.
```
highlighted that this function takes `modin.pandas.DataFrame` objects instead of its query compilers

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -524,285 +1735,616 @@ def set_index_from_columns(
-    # instead.
-    def all(self, **kwargs):
-        """Returns whether all the elements are true, potentially over an axis.
+        .. warning::
```
moved this to the `Notes` section

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -827,143 +2383,356 @@ def describe(self, **kwargs):
+    @add_refer_to("DataFrame.rank")
+    def rank(self, **kwargs):  # noqa: PR02
+        """
+        Get numerical rank for each value along axis.
```
idk actually)
reverted it back

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -972,15 +2741,21 @@ def getitem_array(df, key):
     def getitem_column_array(self, key, numeric=False):
-        """Get column data for target labels.
+        """
+        Get column data for target labels.
```
- [`getitem_array`](https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/backends/pandas/query_compiler.py#L2140) masks query compiler with either a 2D boolean mask or a list of columns.
- [`getitem_column_array`](https://github.com/modin-project/modin/blob/a3ddf2f01163a312416d2a8bc456ba9582ae9b4d/modin/backends/pandas/query_compiler.py#L2177) masks query compiler only with a list of columns.

yeah, it seems that the `getitem_column_array` duplicates some part of the logic of the `getitem_array`, but there is nothing that docs can do about it

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1438,6 +3092,36 @@ def groupby_agg(
+            GroupBy parameters in the format of ``modin.pandas.DataFrame.groupby`` signature.
+        drop : bool, default: False
+            If `by` is a `QueryCompiler` indicates whether or not by-data came
+            from the `self`.
```
[answer](https://github.com/modin-project/modin/pull/3021#discussion_r641743517)

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1478,6 +3191,25 @@ def pivot_table(
+
+        Parameters
+        ----------
+        index : label, pandas.Grouper, array or list of such
+        values : label, optional
+        columns : column, pandas.Grouper, array or list of such
+        aggfunc : callable(pandas.Series) -> scalar, dict of list of such
+        fill_value : scalar, optional
+        margins : bool
+        dropna : bool
+        margins_name : str
+        observed : bool
```
[answer](https://github.com/modin-project/modin/pull/3021#discussion_r641728969)

dchigarev commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+'''''''
+As an exercise let's define a new query compiler in `Modin`, just to see how easy it is.
+Usually, the query compiler routes formed queries to the underlying :doc:`frame </flow/modin/engines/base/frame/data>` class,
+which represents submits operators to an execution engine. For the sake
```
fixed

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -24,15 +31,831 @@
+)
+
+
+def _add_deprecation_warning(replacement_method):
```
moved all decorators to the separate file `doc_utils.py`

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -239,204 +1170,443 @@ def to_numpy(self, **kwargs):
+        other : BaseQueryCompiler or NumPy array
+            The other query compiler or NumPy array to matrix multiply with self.
+        squeeze_self : boolean
+            The flag to squeeze self.
```
changed to the following:
"If `self` is a one-column query compiler, indicates whether it represents Series object."

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1644,6 +3441,25 @@ def setitem(df, axis, key, value):
 
     def write_items(self, row_numeric_index, col_numeric_index, broadcasted_items):
+        """
+        Update QueryCompiler elements at the specified positions by passed values.
```
`write_items` allows doing 2D assignments of non-distributed `broadcasted_items`. Mentioned this in the docs

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+
+        Parameters
+        ----------
+        tz : str, pytz.timezone, optional
```
fixed, this parameter shouldn't be optional

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+            self, resample_args, _method, *args, **kwargs
+        )
+
+    # FIXME: high-level objects leaks to the query compiler.
```
this function takes `modin.pandas.DataFrame` instead of its query compiler, mentioned this in the comment

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+
+    @_doc_dt_timestamp(property="hour", method="hour")
+    def dt_hour(self):
+        return DateTimeDefault.register(pandas.Series.dt.hour)(self)
```
[answer](https://github.com/modin-project/modin/pull/3021#discussion_r641793369)

dchigarev commented on docs/flow/modin/backends/base/query_compiler.rst:
```diff
@@ -0,0 +1,97 @@
+As an exercise let's define a new query compiler in `Modin`, just to see how easy it is.
+Usually, the query compiler routes formed queries to the underlying :doc:`frame </flow/modin/engines/base/frame/data>` class,
+which represents submits operators to an execution engine. For the sake
+of simplicity and independence of this example, our execution engine will be the `pandas` itself.
```
no, removed backticks

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+        return DateTimeDefault.register(pandas.Series.dt.to_pydatetime)(self)
+
+    # FIXME: there are no references to this method, we should either remove it
+    # or add a call reference at the DataFrame level.
```
Created #3103

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +3577,1371 @@ def set_index_names(self, names, axis=0):
+
+    # FIXME:
+    #   1. Backend shouldn't care about differences between Series and DataFrame
+    #      so `resample_agg_df` and `resample_agg_ser` should be combined.
```
Created an issue (#3104) where described the differences between processing DataFrames and Series, and why it should be handled at the front

dchigarev commented:
Merging with master created a merge-commit with the message that doesn't follow our commit-messages style, so the CI is failing now :tired_face:, so I'll have to rebase it anyway after the review is over to rename those commits. 

But anyway, here are the highlights of the changes:
1. Implement [`format_string`](https://github.com/dchigarev/modin/blob/1f64e7e5b6374fc505c05a27ce9b6c59bcf5a85c/modin/utils.py#L101) function which does template parameters substitution with correct indents. This allowed to get rid of all of those places with confusing string indents. Commit:  9ea8b80
2. Applied @vnlitvinov's [suggestion](https://github.com/modin-project/modin/pull/3021#discussion_r627431603) and used `functools.partial` for most of the doc decorators. Commit: 8e2d565
3. Moved doc decorators to the separate file: `modin/backends/base/doc_utils.py`. Commit: ec4a8bb
4. Added tests for newly added functions to `modin/utils.py`. Commits: 308c6ec and 9ea8b80
5. Applied other @devin-petersohn's and @vnlitvinov's suggestions about docs content.

TeamCity CI is green, so this PR is ready for review.

devin-petersohn reviewed (dismissed):
LGTM! Thanks @dchigarev

devin-petersohn commented:
@vnlitvinov You have requested changes, does it look okay to you?

vnlitvinov commented:
> @vnlitvinov You have requested changes, does it look okay to you?

I'm looking into that, the PR is huuuuuuuuge :)

devin-petersohn reviewed (commented):
@dchigarev Can you clear the conflicts?

YarShev commented:
@vnlitvinov , any progress on looking into this? :)

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+    template = """
+    Perform element-wise {operation} (``{verbose}``).
+
+    If axes are note equal, first perform frames allignment.
```
```suggestion
    If axes are not equal, perform frames alignment first.
```

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+    broadcast : bool, default: False
+        If `other` is a one-column query compiler, indicates whether it is a Series or not.
+        Frames and Series have to be processed differently, however we can't distinguish them
+        at the query compiler level, so this parameter is a hint that passed from a high level API.
```
```suggestion
        at the query compiler level, so this parameter is a hint that is passed from a high-level API.
```

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+        level : int or label
+            In case of MultiIndex match index values on the passed level.
+        axis : {{0, 1}}
+            Axis to match indice along for 1D `other` (list or QueryCompiler that represents Series).
```
```suggestion
            Axis to match indices along for 1D `other` (list or QueryCompiler that represents Series).
```

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+        level : int or label
+            In case of MultiIndex match index values on the passed level.
+        axis : {{0, 1}}
+            Axis to match indice along for 1D `other` (list or QueryCompiler that represents Series).
```
```suggestion
            Axis to match indices along for 1D `other` (list or QueryCompiler that represents Series).
```

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+            Axis to match indice along for 1D `other` (list or QueryCompiler that represents Series).
+            0 is for index, when 1 is for columns.
+        fill_value : float or None
+            Value to fill missing elements in the result of frame allignment.
```
```suggestion
            Value to fill missing elements during frame alignment.
```

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+    params = "".join(
+        [params]
+        + [
+            extra_params_map.get(param, f"{param} : object\n")
```
this template for missing arguments looks different, it doesn't include a newline or some whitespace in the beginning of the help line... something's off here.

dchigarev replied:
right, aligned this template to the ones in the `extra_params_map`

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+            Serves the compatibility purpose. Does not affect the result.""",
+    }
+
+    params = "".join(
+        [params]
+        + [
```
```suggestion
    params += "".join(
        [
```
?

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+        Method parameters in the NumPy docstyle format to substitute
+        to the docstring template.
+    compatibility_params : bool, default: True
+        Whether method takes `*args` and `**kwargs` that does not affect
```
```suggestion
        Whether method takes `*args` and `**kwargs` that do not affect
```

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+            )
+        )
+
+    build_rules = "- QueryCompiler contains unsampled data with missing values filled."
```
could you clarify please what this means?

dchigarev replied:
`build_rules` is a string formally describing how the function's output looks like. This one is substitution at the general [resample functions template](https://github.com/dchigarev/modin/blob/016e44f8d54354bdfc75ef8812732807b60ca51e/modin/backends/base/doc_utils.py#L322).

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+            - Output QueryCompiler has the same shape and axes labels as the source.
+            - Each element is the {result} for the corresponding window.""",
+        "udf_aggregation": """
+            - Labels on the specified axis is preserved.
+            - Labels on the opposit of specified axis is MultiIndex, where first level
+              contains preserved labels of this axis and the second level is the function names.
```
```suggestion
            - Labels on the specified axis are preserved.
            - Labels on the opposit of specified axis are MultiIndex, where first level
              contains preserved labels of this axis and the second level has the function names.
```

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+    groupby_args : dict
+        GroupBy parameters as expected by ``modin.pandas.DataFrame.groupby`` signature.
+    map_args : dict
+        Keyword arguments to pass to the reduction function. If GroupBy is implemented with MapReduce
+        approach, passes this arguments at the map phase only.
```
```suggestion
        Keyword arguments to pass to the reduction function. If GroupBy is implemented via MapReduce
        approach, this argument is passed at the map phase only.
```

vnlitvinov commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+        QueryCompiler containing the result of groupby reduction built by the
+        following rules:
+
+        - Labels on the opposit of specified axis is preserved.
```
```suggestion
        - Labels on the opposit of specified axis are preserved.
```

vnlitvinov commented on modin/utils.py:
```diff
@@ -58,15 +64,143 @@ def _get_indent(doc: str) -> int:
+    str
+        Formated string.
+    """
+    # We want to change indentation only for those values, whose placeholder located
+    # at the beginning of the line, in that case, the placeholder sets an indentation
```
```suggestion
    # We want to change indentation only for those values which placeholders are located
    # at the start of the line, in that case the placeholder sets an indentation
```

vnlitvinov commented on modin/utils.py:
```diff
@@ -58,15 +64,143 @@ def _get_indent(doc: str) -> int:
+    # that the filling value has to obey.
+    template_lines = template.splitlines()
+    # RegExp determining placeholders located at the beginning of the line.
+    regex = r"^[ ^{]*{(\w+)}"
```
```suggestion
    regex = r"^ *{(\w+)}"
```
your form of `[ ^{]` is not matching your intent (I think you wanted to match "space but not curly brace", but `^` means "not" in regex _only_ if it's the first symbol after square bracket, otherwise it's just a normal symbol).

vnlitvinov commented on modin/utils.py:
```diff
@@ -58,15 +64,143 @@ def _get_indent(doc: str) -> int:
+        match = re.search(regex, line)
+        if match is None:
+            continue
+        value_key = match.group(1)
```
Why did't you just match against `r"^( *){(\w+)}"`? This way it would have given you _both_ the substitution entity _and_ the indent of it as two match groups... see https://regex101.com/r/8kJR7j/1

dchigarev replied:
changed my regexp with yours

vnlitvinov commented on modin/utils.py:
```diff
@@ -58,15 +64,143 @@ def _get_indent(doc: str) -> int:
+        # Since placeholder is located at the beginning of a new line, its already
+        # has '\n' before it, so to avoid double new lines we want to ensure that
+        # the value string doesn't starts with '\n'
+        if value[0] == "\n":
+            value = value[1:]
```
```suggestion
        value = value.lstrip("\n")
```
?

dchigarev replied:
My explanation at the above comment wasn't clear enough, we don't want to omit all of the leading '\n' characters, otherwise, it'll be impossible to substitute a value with the leading empty line.

<details><summary>Example</summary>

```python
from modin.utils import format_string

value1 = """
    I don't want to have leading '\\n' when this value
    will be substituted
"""

value2 = """

    And here I want to have leading '\\n' because I left
    an explicit empty line at the beginning
"""

template = """
    Placeholder1:
    {placeholder1}
    Placeholder2:
    {placeholder2}
"""

print(format_string(template, placeholder1=value1, placeholder2=value2))

# Output:
#   Placeholder1:
#   I don't want to have leading '\n' when this value
#   will be substituted
#
#   Placeholder2:
#
#   And here I want to have leading '\n' because I left
#   an explicit empty line at the beginning
#

```

</details>

vnlitvinov commented on modin/utils.py:
```diff
@@ -58,15 +64,143 @@ def _get_indent(doc: str) -> int:
+    return indents
+
+
+def format_string(template: str, **kwargs) -> str:
```
this function better have a test...

dchigarev replied:
it already has [one](https://github.com/modin-project/modin/pull/3021/files#diff-003a411f9eba9c58275576d3bfeddd0815d72660fce91c34f77052d6fe5b92a5R185)

vnlitvinov reviewed (commented):
Easy chunk of the review - everything that's _not_ `query_compiler.py`. It will be reviewed in a separate review.

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +2818,1442 @@ def set_index_names(self, names, axis=0):
+            self, resample_args, fill_value
+        )
+
+    # FIXME: `resample_backfill` is an alias for `resample_bfill`, on of these method
```
```suggestion
    # FIXME: `resample_backfill` is an alias for `resample_bfill`, one of these method
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +2818,1442 @@ def set_index_names(self, names, axis=0):
+            self, resample_args
+        )
+
+    # FIXME: `resample_ffill` is an alias for `resample_pad`, on of these method
```
```suggestion
    # FIXME: `resample_ffill` is an alias for `resample_pad`, one of these method
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +2818,1442 @@ def set_index_names(self, names, axis=0):
+            self, resample_args, limit
+        )
+
+    # FIXME: we should combine all method all resample fillna methods into `resample_fillna`
```
```suggestion
    # FIXME: we should combine all resample fillna methods into `resample_fillna`
```
?

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +2818,1442 @@ def set_index_names(self, names, axis=0):
+        refer_to="interpolate",
+        params="""
+        method : str
+        axis : {{0, 1}}
```
```suggestion
        axis : {0, 1}
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +2818,1442 @@ def set_index_names(self, names, axis=0):
+    @doc_utils.doc_reduce_agg(
+        method="mean absolute deviation",
+        params="""
+        axis : {{0, 1}}
```
```suggestion
        axis : {0, 1}
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +2818,1442 @@ def set_index_names(self, names, axis=0):
+        )
+
+    @doc_utils.doc_reduce_agg(
+        method="unibased kurtosis", refer_to="kurt", extra_params=["skipna", "**kwargs"]
```
```suggestion
        method="unbiased kurtosis", refer_to="kurt", extra_params=["skipna", "**kwargs"]
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +2818,1442 @@ def set_index_names(self, names, axis=0):
+        )
+
+    @doc_utils.doc_window_method(
+        result="unibased kurtosis", refer_to="kurt", params="**kwargs : dict"
```
```suggestion
        result="unbiased kurtosis", refer_to="kurt", params="**kwargs : dict"
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1740,218 +2818,1442 @@ def set_index_names(self, names, axis=0):
+        other : BaseQueryCompiler
+            Query compiler to compare with. Have to be the same shape and the same
+            labeling as `self`.
+        align_axis : {{0, 1}}
```
```suggestion
        align_axis : {0, 1}
```

dchigarev commented on modin/test/test_utils.py:
```diff
@@ -129,3 +131,102 @@ class Child(Parent):
+    source = """
+    Source string that sets
+        the indent pattern."""
+    target = indent(source, " " * 5)
```
here I explicitly change indents for the `source` string, the resulted string became the `target` one which we're aligning with the `source` via the `align_indents` function

dchigarev commented on modin/backends/base/doc_utils.py:
```diff
@@ -0,0 +1,685 @@
+        """
+
+    params = format_string(params_template, params=params) if params else ""
+    substituted = format_string(template, params=params, refer_to=refer_to, **kwargs)
+    if refer_to_module_name:
+        refer_to = f"{refer_to_module_name}.{refer_to}"
```
An entity name always like "module_name.entity", isn't it?

vnlitvinov replied:
but you're also using `refer_to` in its short form in line 120, why?

dchigarev replied:
In line 120 we're substituting raw values to the template, they shouldn't be processed anyhow, because when the source template contains `{refer_to}` placeholder we expect that the actual `refer_to` parameter value will be substituted. The actual refer-to note is appended [further](https://github.com/dchigarev/modin/blob/016e44f8d54354bdfc75ef8812732807b60ca51e/modin/backends/base/doc_utils.py#L127-L128) with the proper code entity name.
```python
from modin.backends.base.doc_utils import doc_qc_method

def foo():
    pass

print(
    doc_qc_method(
        template="Perform {refer_to} operation.",
        refer_to="foo",
        refer_to_module_name="bar",
    )(foo).__doc__
)

# Output:
# Perform foo operation.
#
# Notes
# -----
# Please refer to ``modin.pandas.bar.foo`` for more information
# about parameters and output format.
```

vnlitvinov replied:
So this looks like somewhat bad naming, where you're using `refer_to` both for referring to something else and to name the exact operation... however this is really minor, so we can leave it be for now.

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -50,31 +84,49 @@ def axis_setter(self, labels):
-    Note: See the Abstract Methods and Fields section immediately below this
-        for a list of requirements for subclassing this object.
+    """
+    Abstract class that handles the queries to Modin dataframes.
+
+    This class defines common query compilers API, most of the methods
```
This sentence tells that that *most* of the methods are implemented, some of the methods are still being abstract, so the class is abstract as well

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1576,16 +2571,32 @@ def get_axis(self, axis):
 
     def view(self, index=None, columns=None):
+        """
+        Mask QueryCompiler with passed keys.
```
`getitem_array` allows the user to select either rows or columns when via `view` we can pick both

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -155,7 +267,7 @@ def concat(df, axis, other, **kwargs):
     def free(self):
-        """In the future, this will hopefully trigger a cleanup of this object."""
+        """Trigger a cleanup of this object."""
         # TODO create a way to clean up this object.
```
removed

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -239,204 +368,458 @@ def to_numpy(self, **kwargs):
+        upper : float or list-like
+        axis : {{0, 1}}
+        inplace : bool
+            Serves the compatibility purpose. Does not affect the result.
```
actually, this always has to be `False`, changed the doc

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -462,52 +845,94 @@ def columnarize(self):
+    @doc_utils.add_refer_to("DataFrame.reindex")
+    def reindex(self, axis, labels, **kwargs):  # noqa: PR02
+        """
+        Allign QueryCompiler data with a new index-labels along specified axis.
```
did `Allign` -> `Align` and also `index-labels` -> `index`

dchigarev commented on modin/backends/base/query_compiler.py:
```diff
@@ -1476,6 +2381,36 @@ def groupby_agg(
+            Axis to group and apply aggregation function along.
+            0 is for index, when 1 is for columns.
+        agg_func : dict or callable(pandas.core.groupby.DataFrameGroupBy) -> pandas.DataFrame
+            Function to apply to the pandas GroupBy object.
```
changed "pandas Groupby object" to just "GroupBy object"

vnlitvinov commented on modin/utils.py:
```diff
@@ -117,40 +116,40 @@ def format_string(template: str, **kwargs) -> str:
+        # Since placeholder is located at the beginning of a new line,
+        # it already has '\n' before it, so to avoid double new lines
+        # we want to discard the first leading '\n' at the value line,
+        # the others leading '\n' considered as a purposed new lines
```
```suggestion
        # the others leading '\n' are considered as being put on purpose
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1725,7 +1679,7 @@ def nlargest(self, n=5, columns=None, keep="first"):
         columns : list of labels, optional
             Column labels to order by.
-            (note: this parameter can be ommited only for a single-column query compilers
+            (note: this parameter can be omited only for a single-column query compilers
```
```suggestion
            (note: this parameter can be omitted only for a single-column query compilers
```

vnlitvinov commented on modin/backends/base/query_compiler.py:
```diff
@@ -1750,7 +1704,7 @@ def nsmallest(self, n=5, columns=None, keep="first"):
         columns : list of labels, optional
             Column labels to order by.
-            (note: this parameter can be ommited only for a single-column query compilers
+            (note: this parameter can be omited only for a single-column query compilers
```
```suggestion
            (note: this parameter can be omitted only for a single-column query compilers
```

vnlitvinov reviewed (commented):
Please also answer https://github.com/modin-project/modin/pull/3021#discussion_r670529572

dchigarev commented:
@vnlitvinov all of the comments are answered and CI is green

devin-petersohn reviewed (approved):
LGTM! @vnlitvinov any comments?

vnlitvinov reviewed (approved):
LGTM, thanks @dchigarev for bearing with my nitpicking!
